### PR TITLE
fix: show a helpful error when a managed environment has been deleted

### PR DESCRIPTION
**Pull request by nicklewis**
Previously, we showed a generic failure message when the upstream 
environment was deleted. We now handle that case and show a message 
indicating what's happened and instructing the user how to restore the 
environment.

This also changes the way we handle EnvironmentSelectErrors. We now 
include a check for EnvironmentSelectError in main() alongside the other error
handler, and handle the EnvironmentSelectError::Environment variant the same as
ordinary EnvironmentErrors.

### CODE REVIEW

**Comment by nicklewis** [comment]
There are probably some tests to add for this, but here it is for review at least.

**Review by jennymahmoudi — CHANGES_REQUESTED** [review]
For the message itself, can you update it to:

> Run 'flox push' to restore the environment to FloxHub.

Note the single quotes instead of backticks (style guide).

**Comment by nicklewis** [comment]
Updated to use single quotes and to only show the message if you're the owner of the environment.

**Review by jennymahmoudi — APPROVED** [review]

**Review by ysndr — APPROVED** [review]
Error handling is spread rather broadly and apparently we have some duplication to deal with through explicit handling not removed yet or parallel efforts in here and #1632.

Functionally: lgtm
still worth a greenkeeping ticket to deal with the handling duplications.

File: cli/flox/src/main.rs [hunk 1]
```diff
@@ -132,6 +132,13 @@ fn main() -> ExitCode {
                 return e.downcast_ref::<FloxShellErrorCode>().unwrap().0;
             }
 
+            if let Some(EnvironmentSelectError::Environment(e)) =
```
[the remark this answers is not in the document]
Comment (nicklewis) [reply cli/flox/src/main.rs]: Updated

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 2]
```diff
@@ -101,7 +101,7 @@ pub enum ManagedEnvironmentError {
     #[error("access to floxmeta repository was denied")]
     AccessDenied,
     #[error("environment '{0}' does not exist at upstream '{1}'")]
-    UpstreamNotFound(EnvironmentRef, String),
+    UpstreamNotFound(EnvironmentRef, String, Option<String>),
```
Comment (zmitchell) [inline cli/flox-rust-sdk/src/models/environment/managed_environment.rs:104]: Suggestion: make this a struct variant with named fields, it's unclear what the `Option<String>` is supposed to represent from looking at this.

Something like this
```rust
#[error("environment '{name}' does not exist at upstream '{upstream}'")]
UpstreamNotFound {
    name: EnvironmentRef,
    upstream: String,
    whatever_this_one_is: Option<String>,
}
```

**Review by zmitchell — APPROVED** [review]

**Review by zmitchell — APPROVED** [review]

**Pull request merged**